Signal/src/util/ThreadUtil.m

```diff
@@ -84,6 +88,112 @@ + (void)sendMessageWithAttachment:(SignalAttachment *)attachment
         }];
 }
 
++ (void)createBlockOfferIfNecessary:(TSContactThread *)contactThread
+                     storageManager:(TSStorageManager *)storageManager
+                    contactsManager:(OWSContactsManager *)contactsManager
+                    blockingManager:(OWSBlockingManager *)blockingManager
+{
+    OWSAssert(contactThread);
+    OWSAssert(storageManager);
+    OWSAssert(contactsManager);
+    OWSAssert(blockingManager);
```

Comment:
Since we call this method every time we open a 1:1 conversation, performance matters.  I've done what I can to mitigate the perf cost; I intend to measure the cost for worst-case conversations (e.g. conversations with 10k incoming messages and no outgoing messages).